app/src/main/java/org/fcitx/fcitx5/android/input/keyboard/CustomGestureView.kt

```diff
@@ -97,7 +99,35 @@ open class CustomGestureView(ctx: Context) : FrameLayout(ctx) {
 
     private fun hapticFeedback(feedback: Int = HapticFeedbackConstants.KEYBOARD_TAP) {
         if (buttonHapticFeedback) {
-            performHapticFeedback(feedback, HapticFeedbackFlags)
+            if (feedback == HapticFeedbackConstants.KEYBOARD_TAP) {
+                if (buttonPressVibrationMilliseconds==0)
+                    performHapticFeedback(feedback, HapticFeedbackFlags)
+                else if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+                    vibrator.vibrate(VibrationEffect.createOneShot(
+                        buttonPressVibrationMilliseconds.toLong(),
+                        if(vibrator.hasAmplitudeControl())
+                            buttonPressVibrationAmplitude
+                        else
+                            VibrationEffect.DEFAULT_AMPLITUDE))
+                } else {
+                    @Suppress("DEPRECATION")
+                    vibrator.vibrate(buttonPressVibrationMilliseconds.toLong())
+                }
+            } else {
+                if (buttonLongPressVibrationMilliseconds==0)
+                    performHapticFeedback(feedback, HapticFeedbackFlags)
+                else if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+                    vibrator.vibrate(VibrationEffect.createOneShot(
+                        buttonLongPressVibrationMilliseconds.toLong(),
+                        if(vibrator.hasAmplitudeControl())
+                            buttonLongPressVibrationAmplitude
+                        else
+                            VibrationEffect.DEFAULT_AMPLITUDE))
+                } else {
+                    @Suppress("DEPRECATION")
+                    vibrator.vibrate(buttonLongPressVibrationMilliseconds.toLong())
+                }
+            }
```

Comment:
Those conditions could be merged. Or it's better to create a helper class to avoid calling `vibrator.hasAmplitudeControl()` every time.